File: my-folder/0098-validate-binary-search-tree/solution.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public bool IsValidBST(TreeNode root) {
        bool validRangeRecursion(TreeNode node, int? min, int? max) {
            if (node == null) return true;

            if ((min.HasValue && node.val <= min) 
                || (max.HasValue && node.val >= max)) return false;
            
            return validRangeRecursion(node.left, min, node.val)
                && validRangeRecursion(node.right, node.val, max);
        }

        return validRangeRecursion(root, null, null);
    }
}

File: my-folder/0098-validate-binary-search-tree/solution.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isValidBST = function(root) {
    function validRangeRecursion(node, min, max) {
        if (!node) return true;

        if (node.val <= min || node.val >= max) {
            return false;
        }

        return (
        validRangeRecursion(node.left, min, node.val) &&                 
        validRangeRecursion(node.right, node.val, max));
    }

    return validRangeRecursion(root, -Infinity, Infinity);
};
